src/common/hobject.h

```diff
@@ -344,14 +372,11 @@ CEPH_HASH_NAMESPACE_END
 
 ostream& operator<<(ostream& out, const ghobject_t& o);
 
-WRITE_EQ_OPERATORS_3(ghobject_t, hobj, shard_id, generation)
-// sort ghobject_t's by <hobj, shard_id, generation> 
-// 
-// Two objects which differ by generation are more related than
-// two objects of the same generation which differ by shard.
-// 
-WRITE_CMP_OPERATORS_3(ghobject_t,
-		      hobj,
+WRITE_EQ_OPERATORS_4(ghobject_t, max, shard_id, hobj, generation)
+
+WRITE_CMP_OPERATORS_4(ghobject_t,
+		      max,
```

Comment:
Keep max as the most significant field I'm just saying you don't need a 
second "bool max" in the ghobject_t.  As a matter of fact the 
WRITE_EQ_OPERATORS_3() can stay the same since the max will be checked 
in the hobj.

Something like this:

WRITE_EQ_OPERATORS_3(ghobject_t, hobj, shard_id, generation)

WRITE_CMP_OPERATORS_4(ghobject_t,
                hobj.max,
                shard_id,
                hobj,
                generation)


On 1/8/15 4:13 PM, Sage Weil wrote:
>> @@ -344,14 +372,11 @@ CEPH_HASH_NAMESPACE_END
>>   
>>   ostream& operator<<(ostream& out, const ghobject_t& o);
>>   
>> -WRITE_EQ_OPERATORS_3(ghobject_t, hobj, shard_id, generation)
>> -// sort ghobject_t's by <hobj, shard_id, generation>
>> -//
>> -// Two objects which differ by generation are more related than
>> -// two objects of the same generation which differ by shard.
>> -//
>> -WRITE_CMP_OPERATORS_3(ghobject_t,
>> -		      hobj,
>> +WRITE_EQ_OPERATORS_4(ghobject_t, max, shard_id, hobj, generation)
>> +
>> +WRITE_CMP_OPERATORS_4(ghobject_t,
>> +		      max,
> On Thu, 8 Jan 2015, David Zafman wrote:
>> In src/common/hobject.h:
>>
>>> @@ -344,14 +372,11 @@ CEPH_HASH_NAMESPACE_END
>>>   
>>>   ostream& operator<<(ostream& out, const ghobject_t& o);
>>>   
>>> -WRITE_EQ_OPERATORS_3(ghobject_t, hobj, shard_id, generation)
>>> -// sort ghobject_t's by <hobj, shard_id, generation>
>>> -//
>>> -// Two objects which differ by generation are more related than
>>> -// two objects of the same generation which differ by shard.
>>> -//
>>> -WRITE_CMP_OPERATORS_3(ghobject_t,
>>> -		      hobj,
>>> +WRITE_EQ_OPERATORS_4(ghobject_t, max, shard_id, hobj, generation)
>>> +
>>> +WRITE_CMP_OPERATORS_4(ghobject_t,
>>> +		      max,
>> Do WRITE_EQ_OPERATORS_4 and WRITE_CMP_OPERATORS_4 without adding a bool max to ghobject_t. These macros should still work if you have hobj.max with friend class.
> I tried going down this route but ran into problems when max was not
> the most significant field (i.e. when it was not more significant than
> hobject_t::max).  i forget exactly where it when wrong but it was
> annoying.  in any case, making a distinct value seemed cleaner anyway
>
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/ceph/ceph/pull/3143/files#r22693857